cmd/soroban-rpc/internal/xdr2json/conversion.go

```diff
@@ -0,0 +1,139 @@
+
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/db"
+)
+
+const (
+	FormatBase64 = "base64"
+	FormatJSON   = "json"
+)
+
+var errInvalidFormat = fmt.Errorf(
+	"expected %s for optional 'xdrFormat'",
+	strings.Join([]string{FormatBase64, FormatJSON}, ", "))
+
+// ConvertBytes takes an XDR object (`xdr`) and its serialized bytes (`field`)
+// and returns the raw JSON-formatted serialization of that object.
+// It can be unmarshalled to a proper JSON structure, but the raw bytes are
+// returned to avoid unnecessary round-trips. If there is an
+// error, it returns an empty JSON object.
+//
+// The `xdr` object does not need to actually be initialized/valid:
+// we only use it to determine the name of the structure. We could just
+// accept a string, but that would make mistakes likelier than passing the
+// structure itself (by reference).
+func ConvertBytes(xdr interface{}, field []byte) ([]byte, error) {
+	xdrTypeName := reflect.TypeOf(xdr).Name()
+	return convertAnyBytes(xdrTypeName, field)
+}
+
+// ConvertInterface takes a valid XDR object (`xdr`) and returns
+// the raw JSON-formatted serialization of that object. If there is an
+// error, it returns an empty JSON object.
+//
+// Unlike `ConvertBytes`, the value here needs to be valid and
+// serializable.
+func ConvertInterface(xdr encoding.BinaryMarshaler) (json.RawMessage, error) {
+	xdrTypeName := reflect.TypeOf(xdr).Name()
+	data, err := xdr.MarshalBinary()
+	if err != nil {
+		return []byte("{}"), errors.Wrapf(err, "failed to serialize XDR type '%s'", xdrTypeName)
+	}
+
+	return convertAnyBytes(xdrTypeName, data)
+}
+
+func convertAnyBytes(xdrTypeName string, field []byte) (json.RawMessage, error) {
+	var jsonStr, errStr string
+	// scope just added to show matching alloc/frees
+	{
+		goRawXdr := CXDR(field)
+		b := C.CString(xdrTypeName)
+
+		result := C.xdr_to_json(b, goRawXdr)
+		C.free(unsafe.Pointer(b))
+
+		jsonStr = C.GoString(result.json)
+		errStr = C.GoString(result.error)
+
+		C.free_conversion_result(result)
+	}
+
+	if errStr != "" {
+		return json.RawMessage(jsonStr), errors.New(errStr)
+	}
+
+	return json.RawMessage(jsonStr), nil
+}
+
+// CXDR is ripped directly from preflight.go to avoid a dependency.
+func CXDR(xdr []byte) C.xdr_t {
+	return C.xdr_t{
+		xdr: (*C.uchar)(C.CBytes(xdr)),
+		len: C.size_t(len(xdr)),
+	}
+}
+
+func TransactionToJSON(tx db.Transaction) (
+	[]byte,
+	[]byte,
+	[]byte,
+	error,
+) {
+	var err error
+	var result, resultMeta, envelope []byte
+
+	result, err = ConvertBytes(xdr.TransactionResult{}, tx.Result)
+	if err != nil {
+		return result, envelope, resultMeta, err
+	}
+
+	envelope, err = ConvertBytes(xdr.TransactionEnvelope{}, tx.Envelope)
+	if err != nil {
+		return result, envelope, resultMeta, err
+	}
+
+	resultMeta, err = ConvertBytes(xdr.TransactionMeta{}, tx.Meta)
+	if err != nil {
+		return result, envelope, resultMeta, err
+	}
+
+	return result, envelope, resultMeta, nil
+}
+
+func IsValidConversion(format string) error {
```

Comment:
nit: I think `IsValidFormat()` would be a better name

Reply:
also, I think this function, along with the FormatJSON and FormatBase64 constants, should be moved to the methods package. the xdr2json package seems to be solely focused on converting binary xdr to json and it doesn't need to be aware of other formats

Reply:
Yep, agreed :+1: moved to methods and renamed